Feature: Block-based template parts for Classic themes

Mamaduka opened the pull request:
PR backports changes are required for using block-based template parts in classic themes - https://github.com/WordPress/gutenberg/pull/42729.

Gutenberg tracking issue: https://github.com/WordPress/gutenberg/issues/43440
Trac ticket: https://core.trac.wordpress.org/ticket/56467

## Testing Instructions

**Classic theme  - TT1**

* The new "Template Parts" menu shouldn't appear in Appearance.
* Users can't access the `/wp-admin/site-editor.php` path directly.

**Block theme - TT2**

* The new "Template Parts" menu shouldn't appear in Appearance.
* The "Editor" menu is accessible.
* The Site Editor works as before.

**Themes with block-based templates part enabled**
You can grab the testing theme from this repo - https://github.com/Mamaduka/block-fragments

* The new "Template Parts" menu is visible under Appearance.
* The menu redirects to the "Template Parts" list view.
* The "Site" and "Templates" navigation items aren't visible.

> **Warning**
> The last item will require the latest package update PR to be merged, and it might not be possible to test immediately.


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Mamaduka committed:
Feature: Block-based template parts for Classic themes

Mamaduka commented on src/wp-admin/site-editor.php:
```diff
@@ -64,14 +64,24 @@ static function( $classes ) {
+	'styles'                    => get_block_editor_theme_styles(),
+	'defaultTemplateTypes'      => $indexed_template_types,
+	'defaultTemplatePartAreas'  => get_allowed_block_template_part_areas(),
+	'supportsLayout'            => WP_Theme_JSON_Resolver::theme_has_support(),
```
Before, the Site Editor always assumed layout support.

Mamaduka commented on src/wp-includes/functions.php:
```diff
@@ -5261,7 +5261,7 @@ function wp_widgets_add_menu() {
 
 	$menu_name = __( 'Widgets' );
-	if ( wp_is_block_theme() ) {
+	if ( wp_is_block_theme() || current_theme_supports( 'block-template-parts' ) ) {
```
We need this to avoid conflicts between the `Customize` and `Widgets` menu positions.

Mamaduka committed:
Update unit tests

ockham reviewed (approved):
Thanks a lot @Mamaduka! LGTM :shipit:

hellofromtonya commented:
@ockham what is the status of the Warning in the description?
>The last item will require the latest package update PR to be merged, and it might not be possible to test immediately.

ockham commented:
> @ockham what is the status of the Warning in the description?
> 
> > The last item will require the latest package update PR to be merged, and it might not be possible to test immediately.

Ah, my bad, I should've elaborated on that. I was indeed able to test and confirm everything except for

> The "Site" and "Templates" navigation items aren't visible.

As @Mamaduka said, we'll only be able to verify this once we've synced `@wordpress/` packages. That [sync](https://github.com/WordPress/wordpress-develop/pull/3154) is still blocked. However, I don't think we need to hold off until that package sync, since part of the logic responsible for conditionally showing the "Site" and "Templates" menu items lives in JS (and can be considered part of those packages, rather than the PHP that this PR touches).

In any case, we can easily test this separately once we sync packages.

Does that sound okay?

hellofromtonya commented:
Testing Report

Env:
- WordPress: `trunk`
- Plugins: none
- Browser: Firefox
- Localhost: wp-env

Results:

Classic theme - TT1
- The new "Template Parts" menu shouldn't appear in Appearance ✅ 
- Users can't access the /wp-admin/site-editor.php path directly ✅ 
<img width="1518" alt="Screen Shot 2022-09-14 at 11 55 15 AM" src="https://user-images.githubusercontent.com/7284611/190216065-ad9da9fc-f7eb-41da-8d8f-1d24f72887c2.png">

Block theme - TT2
- The new "Template Parts" menu shouldn't appear in Appearance ✅ 
- The "Editor" menu is accessible ✅ 
- The Site Editor works as before ✅ 

Themes with block-based templates part enabled
Theme:  https://github.com/Mamaduka/block-fragments

- The new "Template Parts" menu is visible under Appearance ✅ 
- The menu redirects to the "Template Parts" list view ✅ 
- The "Site" and "Templates" navigation items aren't visible ❌ (needs the JS packages updated - per notes in the description)

<img width="650" alt="Screen Shot 2022-09-14 at 11 52 36 AM" src="https://user-images.githubusercontent.com/7284611/190215448-ceacb484-1898-4f84-85a0-6b522d4e5b28.png">

Observations:

- With the block-fragments (classic theme with block-based templates) enabled, I am able to directly access `/wp-admin/site-editor.php` but it's a white screen. No errors in the browser's console or error logs. 

<img width="1587" alt="Screen Shot 2022-09-14 at 11 53 34 AM" src="https://user-images.githubusercontent.com/7284611/190215705-1b2bd9e4-e90e-4d1a-a636-48ea4c20f3b5.png">

@Mamaduka @ockham Is this something that will be fixed with another backport? Or a known issue?

fabiankaegy commented on src/wp-admin/site-editor.php:
```diff
@@ -19,7 +19,7 @@
+if ( ! ( current_theme_supports( 'block-template-parts' ) || wp_is_block_theme() ) ) {
 	wp_die( __( 'The theme you are currently using is not compatible with Full Site Editing.' ) );
 }
 
```
@hellofromtonya I belive something like this would solve the case you uncovered in your testing here: https://github.com/WordPress/wordpress-develop/pull/3234#issuecomment-1247048286

> With the block-fragments (classic theme with block-based templates) enabled, I am able to directly access /wp-admin/site-editor.php but it's a white screen. No errors in the browser's console or error logs.

```suggestion
$is_template_part_editor = isset( $_GET['postType'] ) && 'wp_template_part' === sanitize_key( $_GET['postType'] );
if ( ! wp_is_block_theme() && ! $is_template_part_editor ) {
	wp_die( __( 'The theme you are currently using is not compatible with the Site Editor.' ) );
}

```

As far as I'm aware that was not a known issue and no patch for it had been created. 

@Mamaduka asked me to keep an eye on this PR since he is out traveling. I will create a PR on the Gutenberg Repo to fix the issue there aswell.

fabiankaegy replied:
Actually in looking at the code in Gutenberg it was handled as a completely separate admin page. So this issue only exists in this PR and only needs to get fixed here. No backport to Gutenberg is needed. 

https://github.com/WordPress/gutenberg/blob/d19258cb5b85cea52e85a3e86c9550dd48d44fca/lib/compat/wordpress-6.1/template-parts-screen.php#L5

fabiankaegy replied:
In order to make it easier to test, I have also created a separate PR against @Mamaduka's branch which applies the same fix as the one I have shown in the suggestion above: https://github.com/Mamaduka/wordpress-develop/pull/48

hellofromtonya replied:
Yes, that does resolve it. Well done @fabiankaegy - thank you!

Mamaduka replied:
Thank you!

fabiankaegy committed:
fix prevent unwanted access of site-editor.php

non block themes that support block based template parts should only be allowed to access the site-editor.php file if the query parameter to access the template part editor is set correctly

Mamaduka committed:
Merge pull request #48 from fabiankaegy/backport/site-editor-template-parts-mode-fix

fix prevent unwanted access of site-editor.php

ockham commented on tests/phpunit/tests/rest-api/rest-themes-controller.php:
```diff
@@ -406,7 +407,7 @@ public function test_get_item_schema() {
 		$this->assertArrayHasKey( 'title-tag', $theme_supports );
 		$this->assertArrayHasKey( 'wp-block-styles', $theme_supports );
-		$this->assertCount( 21, $theme_supports );
+		$this->assertCount( 22, $theme_supports );
```
This was bumped to `22` by 72380e31ca0bc05dca5617357412ab8fd346a583 (https://github.com/WordPress/wordpress-develop/pull/3205):

https://github.com/WordPress/wordpress-develop/blob/0b2f05a9f61c4032d0ef3138973b10adee8a032f/tests/phpunit/tests/rest-api/rest-themes-controller.php#L410

So we'll need to rebase and increment it again:

```suggestion
		$this->assertCount( 23, $theme_supports );
```

ockham commented:
Thanks a lot, @fabiankaegy! Looks like this PR is good to go, except for failing unit tests, which I think are due to this: https://github.com/WordPress/wordpress-develop/pull/3234/files#r971773508.

@Mamaduka If you're around, could you rebase and change that number to `23`? 🙏

ockham commented:
@Mamaduka is currently AFK -- I've forked this PR and applied my fix: https://github.com/WordPress/wordpress-develop/pull/3258

hellofromtonya commented:
Thank you everyone for your contributions! Committed PR 3258 (finalization of this PR) via https://core.trac.wordpress.org/ticket/56467.
